src/enzyme_ad/jax/Passes/LLVMToMemrefAccess.cpp

```diff
@@ -0,0 +1,174 @@
+#include "Passes.h"
+
+#include "mlir/Analysis/TopologicalSortUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "src/enzyme_ad/jax/Dialect/Ops.h"
+
+#define DEBUG_TYPE "llvm-to-memref-access"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_LLVMTOMEMREFACCESSPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+
+using PtrVal = TypedValue<LLVM::LLVMPointerType>;
+
+namespace mlir {
+MemRefType recoverMemRefTypeFromKernelCall(enzymexla::KernelCallOp kernelCallOp,
+                                           unsigned argIndex) {
+  auto tensorType =
+      kernelCallOp.getInputs()[argIndex].getType().dyn_cast<RankedTensorType>();
+  if (!tensorType) {
+    return nullptr;
+  }
+  // Convert tensor type to memref type
+  return MemRefType::get(tensorType.getShape(), tensorType.getElementType());
+}
+
+} // namespace mlir
+
+struct LLVMToMemrefAccessPass
+    : public enzyme::impl::LLVMToMemrefAccessPassBase<LLVMToMemrefAccessPass> {
+  using LLVMToMemrefAccessPassBase::LLVMToMemrefAccessPassBase;
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto *ctx = moduleOp->getContext();
+    OpBuilder builder(ctx);
+    SymbolTable symTable(moduleOp);
+
+    // Find all kernels and their enzymexla.kernel_call and enzymexla.jit_call
+    // callers
+    DenseMap<FunctionOpInterface, SetVector<CallOpInterface>> funcToKernelMap;
+    moduleOp->walk([&](CallOpInterface callOp) {
+      auto symbolName =
+          dyn_cast_or_null<SymbolRefAttr>(callOp.getCallableForCallee());
```

Comment:
this shouldn't be any call interface, but just jit_call and kernel_call. We can still store as callopinterface, but we should only walk the Jit and kernel call ops